src/Voron/Platform/Pal.cs

```
@@ -75,7 +73,74 @@ static Pal()
             Int32 errorCode,
             void* sb,
             Int32 capacity,
-            out Int32 specialErrnoCodes
-            );
+            out Int32 specialErrnoCodes);
+
+        [DllImport(LIBRVNPAL, SetLastError = true)]
+        public static extern Int32 rvn_create_and_mmap64_file(
+            string filename,
+            Int64 initialFileSize,
+            PalFlags.MMAP_OPTIONS flags,
+            out void* handle,
+            out void* baseAddress,
+            out Int64 actualFileSize,
+            out Int32 errorCode);
+
+        [DllImport(LIBRVNPAL, SetLastError = true)]
+        public static extern Int32 rvn_prefetch_virtual_memory(
+            void *virtualAddress,
+            Int64 length,
+            out Int32 errorCode);
+
+        [DllImport(LIBRVNPAL, SetLastError = true)]
+        private static extern Int32 rvn_get_system_information(
+            out SystemInformation systemInformation,
+            out Int32 errorCode);
+
+        [DllImport(LIBRVNPAL, SetLastError = true)]
+        public static extern Int32 rvn_memory_sync(
+            void *address,
+            Int64 size,
+            PalFlags.MSYNC_OPTIONS flags,
+            out Int32 errorCode);
+
+        [DllImport(LIBRVNPAL, SetLastError = true)]
+        public static extern Int32 rvn_dispose_handle(
+            string filepath,
+            void* handle,
+            Int32 deleteOnClose,
+            out Int32 errorCodeClose,
```

Only one error code should be returned, you can check the return value to check what was the actual error.

---

but we might have them both fail and we want to log both of them (to understand later if and why we have leftovers)

---

That doesn't matter, we fail early and return the earlier failure code.
Note that we still want to _close_ the file, but if we are already in error mode, we don't care if that failed.

---

The reasoning behind this is that we can't expose our internal state and behavior outside. That changes over time.
What we want to do is to have a common error handling interface.
Return code, `int32_t` with our own error code.
`detailed_error_code` with the system error code.

Any additional data you want, must be encoded in the return code, nothing else.